test/functional/invoke.cpp

```diff
@@ -0,0 +1,261 @@
+/*
+  SLB.Functional
+
+  Copyright Michael Park, 2017
+  Copyright Agustin Berge, 2017
+
+  Distributed under the Boost Software License, Version 1.0.
+  (See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
+*/
+
+#include <slb/functional.hpp>
+
+#include <type_traits>
+#include <utility>
+
+#define CATCH_CONFIG_MAIN
+#include "../catch.hpp"
+
+#define CHECK_NESTED(...)                                                      \
+  do {                                                                         \
+    INFO(__FILE__ "(" << __LINE__ << "): " #__VA_ARGS__);                      \
+    check_##__VA_ARGS__;                                                       \
+  } while (false)
+
+// [func.invoke], invoke
+
+// template <class F, class... Args>
+//   invoke_result_t<F, Args...> invoke(F&& f, Args&&... args)
+//     noexcept(is_nothrow_invocable_v<F, Args...>);
+
+// Account for P0012: "Make exception specifications be part of the type
+// system".
+static const bool p0012 = !std::is_same<void(), void() noexcept>::value;
+
+struct C {
+  int obj;
+  C(int val = 0) : obj(val) {}
+
+  int fun(int base) noexcept(p0012) { return base + 0; }
+  int cfun(int base) const noexcept(p0012) { return base + 1; }
+  int lfun(int base) & noexcept(p0012) { return base + 2; }
+  int rfun(int base) && noexcept(p0012) { return base + 3; }
+  int clfun(int base) const& noexcept(p0012) { return base + 4; }
+  int crfun(int base) const&& noexcept(p0012) { return base + 5; }
+};
+
+struct D : C {
+  D(int val = 0) : C(val) {}
+};
+
+template <typename T, bool IsNothrow = true>
+struct smart_ptr {
+  T* ptr;
+  smart_ptr(T* ptr) : ptr(ptr) {}
+  T& operator*() const noexcept(IsNothrow) { return *ptr; }
+  T* operator->() const noexcept { return ptr; }
```

Comment:
Should the `noexcept` here also be `noexcept(IsNothrow)`?

Reply:
Removed.